docs: rewrite readme (#300)
limebit/medmodels: README.md

<div align="center">

>   <img alt="MedModels Logo" src="https://raw.githubusercontent.com/limebit/medmodels-static/refs/heads/main/logos/logo_black.svg">

Reviewer: It would probably be nice for dark mode if this logo also had a white background. Probably just a white rectangle in the background with rounded corners

Reviewer: but no hard feelings on this

Author: Solved by using the `<picture>`html tag: 

```html
<div align="center">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://raw.githubusercontent.com/limebit/medmodels-static/refs/heads/main/logos/logo_white.svg">
    <source media="(prefers-color-scheme: light)" srcset="https://raw.githubusercontent.com/limebit/medmodels-static/refs/heads/main/logos/logo_black.svg">
    <img alt="MedModels Logo">
  </picture>
</div>
```

Still has a transparent bg though

Reviewer: this is even better than the white background imo

  <a href="https://github.com/limebit/medmodels/actions/workflows/testing.yml">
    <img src="https://github.com/limebit/medmodels/actions/workflows/testing.yml/badge.svg?branch=main" alt="Tests">
  </a>
  <a href="https://pypi.org/project/medmodels/">
    <img src="https://img.shields.io/pypi/v/medmodels" alt="PyPI Version">
  </a>
</div>

<h2 align="center">
  MedModels: A Rust powered Python Framework for Modern Healthcare Research
</h2>

## Motivation

> Analyzing real-world evidence, especially patient data, is a complex task demanding accuracy and reproducibility.  Currently, research teams often re-implement the same statistical methods and data processing pipelines, leading to:

Reviewer 2: We need consistency, either real world evidence or real-world evidence

Author: there should not be "real world" without a hyphen in the new readme. There is one "Real World Evidence" in the zenodo.json file. But we need to check out how these keyword values are interpreted by zenodo at first before making a final decision on that

Reviewer 2: Apparently there are more results in Zenodo under Real world evidence than Real-world evidence. So lets keep it that way

* **Inefficient codebases:** Duplication of effort and potential inconsistencies.
* **Faulty implementations:** Increased risk of errors in custom code.

> * **Technical debt:**  Maintaining and updating numerous, disparate codebases becomes challenging.

Reviewer: As this is basically on the top of the readme and also very pronounced because of being bold, this feels like it's describing our library and not the problems we want to solve

Reviewer: also no hard feelings on this. Just what it looked like to me on first glance

Reviewer 2: Agreeable

Author: I see your point, thought about it too. Should we just move that section under the seaction explaining the awesomeness?

Reviewer: We could just change the whole paragraph to something like
```
Analyzing real-world evidence, especially patient data, is a complex task demanding accuracy and reproducibility. Currently, research teams often re-implement the same statistical methods and data processing pipelines, leading to inefficient codebases, faulty implementations and technical debt.
```
Imo we don't need the listing here.

> * **Technical debt:**  Maintaining and updating numerous, disparate codebases becomes challenging.

Reviewer 2: "disparate" sounds weird to me (even though it appears to be right)

Reviewer: Agree

MedModels addresses these challenges by providing a standardized, reliable, and efficient framework for handling, processing, and analyzing electronic health records (EHR) and claims data.

**Target Audience:**

MedModels is designed for a wide range of users working with real-world data and electronic health records, including:

* (Pharmaco-)Epidemiologists
* Real-World Data Analysts
* Health Economists
* Clinicians

> * Data scientists
> * Software developers

Reviewer: ```suggestion
* Data Scientists
* Software Developers
```
should both be capitalized imo if the others are as well

## Key Features

> * **Rust-based Data Class:** Enables efficient transformation of patient data into flexible and extensible network graph structures.

Reviewer: If this is title case, then based should also be capitalized
```suggestion
* **Rust-Based Data Class:** Enables efficient transformation of patient data into flexible and extensible network graph structures.
```

Reviewer: but might be wrong

> * **Rust-based Data Class:** Enables efficient transformation of patient data into flexible and extensible network graph structures.

Reviewer 2: ```suggestion
* **Rust-based Data Class: Facilitates the efficient transformation of patient data into adaptable and scalable network graph structures.
```

> * **High-Performance Computing:**  Handle large datasets in memory while maintaining fast processing speeds due to the underlying Rust implementation.

Reviewer 2: ```suggestion
* **High-Performance Computing:**  Handles large datasets in memory while maintaining fast processing speeds due to the underlying Rust implementation.
```

> * **Standardized Workflows:** Streamline common tasks in real-world evidence analysis, reducing the need for custom code.

Reviewer 2: ```suggestion
* **Standardized Workflows:** Streamlines common tasks in real-world evidence analysis, reducing the need for custom code.
```

> * **Interoperability:**  Facilitates collaboration and data sharing by providing a common data structure and analysis framework.

Reviewer 2: ```suggestion
* **Interoperability:**  Supports collaboration and data sharing through a unified data structure and analysis framework.
```

## Key Components

* **MedRecord Data Structure:**

>     * **Graph-based Representation:**  Organizes medical data using nodes (e.g., patients, medications, diagnoses) and edges (e.g., date, dosage, duration) to capture complex interactions and dependencies.

Reviewer: see comment above

    * **Dynamic Management:**  Provides methods to add, remove, and modify nodes and edges, as well as their associated attributes, allowing for flexible data manipulation.
    * **Efficient Querying:**  Enables efficient querying and retrieval of information from the graph structure, supporting various analytical tasks.
    * **Grouping and Filtering:**  Allows grouping of nodes and edges for simplified management and targeted analysis of specific subsets of data.
    * **High-Performance Backend:**  Built on a Rust backend for optimal performance and efficient handling of large-scale medical datasets.

>     * **Easy Sharing and Storage:** MedRecords can be exported as `.ron` files (Rusty Object Notation), a human-readable format that is easy to share, store, and version control.

Reviewer: wuerde ich nicht als extra punkt aufnehmen momentan. Kann sein, dass sich das vielleicht bald aendert.

Author: welchen Punkt genau? Efficient Querying?

Reviewer: Easy sharing and storage meine ich

Author: it's resolved :)

>     * **Easy Sharing and Storage:** MedRecords can be exported as `.ron` files (Rusty Object Notation), a human-readable format that is easy to share, store, and version control.

Reviewer: Would be nice to include that it is relatively simple to create a medrecrod, e.g. from pandas, polars or default python stuff

Reviewer 2: agreed

Author: section added

>  * **Query Engine**:

Reviewer: already included in the MedRecord data structure part

Author: Yeah I intended to put extra stress on that feature. We could drop it as a seperate section, we also could drop the bullet point under "MedRecord Data Structure"

Reviewer: I think the separate section should be dropped and the bullet point in `MedRecord Data Structure` should be moved up to "second place", directly under **Graph-Based Representation** and above **Dynamic Management**

    * **Flexible Querying:**  Efficiently query nodes and edges of the MedRecord based on various criteria, including attributes, labels, and relationships.
    * **High-Performance Implementation:**  Leverages a Rust backend for fast and efficient query execution, even on large datasets.

* **Treatment Effect Analysis:**

    *  **Estimating Treatment Effects:**  Provides a range of methods for estimating treatment effects from observational data, including:

>         *  **Continuous Outcomes:** Analyze treatment effects on continuous outcomes

Reviewer 2: ```suggestion
        *  **Continuous Outcomes:** Analyze treatment effects on continuous outcomes.
```

        *  **Binary Outcomes:**  Estimate odds ratios, risk ratios, and other metrics for binary outcomes.
        *  **Time-to-Event Outcomes:**  Perform survival analysis and estimate hazard ratios for time-to-event outcomes.
        *  **Effect Size Metrics:** Calculate standardized effect size metrics like Cohen's d and Hedges' g.

    * **Matching:**
        * **(High Dimensional) Propensity Score Matching:**  Reduce confounding bias by matching treated and untreated individuals based on their propensity scores.
        * **Nearest Neighbor Matching:**  Match individuals based on similarity in their observed characteristics.

## Getting Started

**Installation:**

> MedModels can be installed from PyPI using the `pip` command:

Reviewer: maybe add a link somewhere to https://pypi.org/project/medmodels/

Author: "PyPI" now is linking to the PyPi entry of medmodels. Also, the PyPI Version badge on top is linking there too

```python
import pandas as pd
import medmodels as mm

# Patients DataFrame (Nodes)
patients = pd.DataFrame(
    [
        ["Patient 01", 72, "M", "USA"],
        ["Patient 02", 74, "M", "USA"],
        ["Patient 03", 64, "F", "GER"],
    ],
    columns=["ID", "Age", "Sex", "Loc"],
)

# Medications DataFrame (Nodes)
medications = pd.DataFrame(
    [["Med 01", "Insulin"], ["Med 02", "Warfarin"]], columns=["ID", "Name"]
)

# Patients-Medication Relation (Edges)
patient_medication = pd.DataFrame(
    [
        ["Patient 02", "Med 01", pd.Timestamp("20200607")],
        ["Patient 02", "Med 02", pd.Timestamp("20180202")],
        ["Patient 03", "Med 02", pd.Timestamp("20190302")],
    ],
    columns=["Pat_ID", "Med_ID", "Date"],
)

# Create a MedRecord object using the builder pattern
record = mm.MedRecord.builder().add_nodes((patients, "ID"), group="Patients").build()

# Add medications as nodes to the MedRecord
record.add_nodes((medications, "ID"), group="Medications")

# Add edges representing the patient-medication relationships
record.add_edges((patient_medication, "Pat_ID", "Med_ID"))

# Create a group of US patients

> record.add_group("US-Patients", ["Patient 01", "Patient 02"])

Reviewer: this feels a bit weird. all of this should be done during the building phase of the MedRecordBuilder

Reviewer: except if you want to show that you can also add nodes and stuff after the medrecord has been created but then that should be explicitly stated

Author: updated

# Print an overview of the nodes and edges in the MedRecord
record.overview_nodes()

> record.overview_edges()

Reviewer: `print(record)` would show both

## Getting Started

**Installation:**

MedModels can be installed from [PyPI](https://pypi.org/project/medmodels/) using the `pip` command:

```bash
pip install medmodels
```

**Quick Start:**

Here's a quick start guide showing an example of how to use MedModels to create a `MedRecord` object, add nodes and edges, and perform basic operations.

```python
import pandas as pd
import medmodels as mm

# Patients DataFrame (Nodes)
patients = pd.DataFrame(
    [
        ["Patient 01", 72, "M", "USA"],
        ["Patient 02", 74, "M", "USA"],
        ["Patient 03", 64, "F", "GER"],
    ],
    columns=["ID", "Age", "Sex", "Loc"],
)

# Medications DataFrame (Nodes)
medications = pd.DataFrame(
    [["Med 01", "Insulin"], ["Med 02", "Warfarin"]], columns=["ID", "Name"]
)

# Patients-Medication Relation (Edges)
patient_medication = pd.DataFrame(
    [
        ["Patient 02", "Med 01", pd.Timestamp("20200607")],
        ["Patient 02", "Med 02", pd.Timestamp("20180202")],
        ["Patient 03", "Med 02", pd.Timestamp("20190302")],
    ],
    columns=["Pat_ID", "Med_ID", "Date"],
)

# Create a MedRecord object using the builder pattern

> record = mm.MedRecord.builder().add_nodes((patients, "ID"), group="Patients").build()

Reviewer: I didn't mean you should completely remove the other stuff, just add it to the builder :D
Now it wouldn't make sense why the `medications` and `patient_medication` dataframes exist.

I meant:
```suggestion
record = (
    mm.MedRecord.builder()
    .add_nodes((patients, "ID"), group="Patients")
    .add_nodes((medications, "ID"), group="Medications")
    .add_edges((patient_medication, "Pat_ID", "Med_ID"))
    .add_group("US-Patients", nodes=["Patient 01", "Patient 02"])
    .build()
)
```

# Print an combined overview of the nodes and edges in the MedRecord
print(record)

# You can also print only nodes and edges respectively
record.overview_nodes()

> record.overview_edges()

Reviewer: both of these would also need to be wrapped in a `print` afaik

Reviewer 2: Yes, if not they do not show anything (unless it is the end of an executed cell)
